File: deck/base/code/native/node/runtime/channel.ts
// Text channel runtime: an unbounded async queue. send hands the value to a waiting receiver or buffers it; receive
// takes a buffered value or waits for the next send. The opaque handle a seed channel holds is this queue object.
// Reached only through the public channel API.
type SeedChannel = {
  queue: string[]
  waiters: ((value: string) => void)[]
}

const channel = {
  make: (): SeedChannel => ({ queue: [], waiters: [] }),

  send: (target: SeedChannel, item: string): Promise<void> => {
    const waiter = target.waiters.shift()
    if (waiter) waiter(item)
    else target.queue.push(item)
    return Promise.resolve()
  },

  receive: (source: SeedChannel): Promise<string> =>
    new Promise(ok => {
      const value = source.queue.shift()
      if (value !== undefined) ok(value)
      else source.waiters.push(ok)
    }),

  close: (_target: SeedChannel): Promise<void> => Promise.resolve(),
}

File: deck/base/code/native/swift/runtime/channel.swift
// Text channel runtime: a thread-safe buffer with a counting semaphore. send appends and signals; receive waits on the
// semaphore then removes the head, so a value sent before receive is buffered and a receive before send blocks until a
// value arrives. The opaque handle a seed channel holds is this SeedChannel. Reached only through the public channel API.
import Foundation

final class SeedChannel {
    private var buffer: [String] = []
    private let lock = NSLock()
    private let ready = DispatchSemaphore(value: 0)

    func send(_ item: String) {
        lock.lock()
        buffer.append(item)
        lock.unlock()
        ready.signal()
    }

    func receive() -> String {
        ready.wait()
        lock.lock()
        let value = buffer.removeFirst()
        lock.unlock()
        return value
    }
}

enum channel {
    static func make() -> SeedChannel { SeedChannel() }
    static func send(_ target: SeedChannel, _ item: String) async { target.send(item) }
    static func receive(_ source: SeedChannel) async -> String { source.receive() }
    static func close(_ target: SeedChannel) async {}
}
